src/App.tsx

```diff
@@ -1,26 +1,212 @@
+
+    return () => clearTimeout(timeout);
+  }, [error]);
+
+  const titleField = useRef<HTMLInputElement>(null);
+  const [title, setTitle] = useState('');
+
+  useEffect(() => {
+    if (titleField.current) {
+      titleField.current?.focus();
+    }
+  }, [todos]);
+
+  const handleAddTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const trimTitle = title.trim();
+
+    if (!trimTitle) {
+      setError(Error.EmptyTodoTitle);
+
+      return;
+    }
+
+    if (titleField.current) {
+      titleField.current.disabled = true;
+    }
+
+    const newTodo: Omit<Todo, 'id'> = {
+      title: trimTitle,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo({
+      id: 0,
+      ...newTodo,
+    });
+
+    setLoadingTodos(current => [...current, 0]);
+
+    makeTodo(newTodo)
+      .then(newTodoFromServer => {
+        setTodos(currentTodos => [...currentTodos, newTodoFromServer]);
+        setTitle('');
+      })
+      .catch(() => setError(Error.UnableAddTodo))
+      .finally(() => {
+        if (titleField.current) {
+          titleField.current.disabled = false;
+          titleField.current.focus();
+        }
+
+        setTempTodo(null);
+        setLoadingTodos(current => current.filter(todoId => todoId !== 0));
+      });
+  };
+
+  const handleUpdateTodo = (updatedTodo: Todo) => {
+    setLoadingTodos(current => [...current, updatedTodo.id]);
+
+    setTodos(currentTodos => {
+      const newTodos = [...currentTodos];
+      const index = newTodos.findIndex(t => t.id === updatedTodo.id);
+
+      if (newTodos[index].title === updatedTodo.title) {
+        setLoadingTodos(current =>
+          current.filter(todoId => todoId !== updatedTodo.id),
+        );
+
+        return currentTodos;
+      }
+
+      updateTodo(updatedTodo)
+        .then(todo => {
+          newTodos.splice(index, 1, todo);
+
+          return newTodos;
+        })
+        .catch(() => setError(Error.UnableUpdateTodo))
+        .finally(() =>
+          setLoadingTodos(current =>
+            current.filter(todoId => todoId !== updatedTodo.id),
+          ),
+        );
+
+      return newTodos;
+    });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingTodos(current => [...current, todoId]);
+
+    deleteTodo(todoId)
+      .then(() =>
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        ),
+      )
+      .catch(() => setError(Error.UnableDeleteTodo))
+      .finally(() => {
+        setLoadingTodos(current =>
+          current.filter(removedTodoId => todoId !== removedTodoId),
+        );
+      });
+  };
+
+  const deleteAllCompleted = () => {
+    const completedTodoIds = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setLoadingTodos(current => [...current, ...completedTodoIds]);
+
+    const requests = completedTodoIds.map(todoId => deleteTodo(todoId));
+
+    Promise.all(requests)
```

Comment:
Problem with this failed tests looks like here.
do not use Promise.all. tests expects delete todo with success response and not delete with fail.
<img width="629" alt="Screenshot 2024-07-19 at 21 39 50" src="https://github.com/user-attachments/assets/0d9c2392-c464-455a-aeee-306cf3f4129c">